### ISSUE TITLE
SetActiveProgram

### ISSUE THREAD

**Issue by dr-boss** [thetrueavatar/Viessmann-Api#22]
Can You will add SetActiveProgram(), the ActiveEco... ActiveComfort didnt work
maybe the problem are in features.zip file:
 In file heating.circuits.1.operating.programs.active.siren.json the section

` "actions": [
    {
      "isExecutable": true,
      "fields": [
        {
          "enum": [
            "eco",
            "compofrt"
          ],
          "type": "string",
          "name": "program",
          "required": true
        }
      ],
      "href": "http://0.0.0.0:8082/operational-data/installations/291/gateways/7123456789123456/devices/0/features/heating.circuits.0.operating.
      "method": "POST",
      "title": "setProgram",
      "name": "setProgram"
    }
  ],
`
like in  heating.circuits.0.operating.modes.active.siren.json exist for change mode?

and one small problem:
vendor/symfony/polyfill-mbstring/bootstrap.php
Warning: array_walk() expects parameter 1 to be array, string given in phar:///tmp/mnt/ASUS_SYS/entware/share/nginx/html/Viessmann/phar-builder.phar/app/PharBuilder.php on line 387

**Comment by thetrueavatar** [comment]
I'm sorry but I don't quite understand what you mean. features.zip is a sample of file I have found embedded in the ViCare app(dev probably did a mistake by embedding those file).  This helped me to discover functionnalities that my own heating system doesn't support. SetActiveProgram, ActivateEcho,... works on my computer. Could you please tell me which are the steps you're doing so I could try to reproduce the problems ?  Which version of the api ? Error message ?
Well for the warning I'm not an expert on php so I didn't take the time to fix it

**Comment by dr-boss** [comment]
I use for test your last version with php examples.
after:  ActivateComfortProgram.php
I received: 
`Message:
         Unable to set data for featureheating.circuits.0.operating.programs.comfort and action activate and data{}
         Reason: DEVICE_COMMUNICATION_ERROR
Code: 1
` 
I added supplementary function setActiveProgram:
`const OPERATING_PROGRAM = "operating.programs.active";
...
    public function setActiveProgram($program, $circuitId = NULL)
    {
        $this->setRawJsonData($this->buildFeature($circuitId, self::OPERATING_PROGRAMS), 
 "setProgram", "{\"program\":\"" . $program . "\"
    }`
I was the same negative result after call $viessmannApi->setActiveProgram("Eco"); or "Comfort"

Second problem, I can't registry new schedule, after post 
$schedule = "{\"mon\":[{\"start\": \"06:20\",\"end\": \"07:30\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"19:00\",\"end\": \"21:30\",\"mode\": \"on\",\"position\":2}],\"tue\":[{\"start\": \"06:20\",\"end\": \"07:30\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"19:00\",\"end\": \"20:30\",\"mode\": \"on\",\"position\":2}],\"wed\":[{\"start\": \"06:20\",\"end\": \"07:30\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"19:00\",\"end\": \"20:30\",\"mode\": \"on\",\"position\":2}],\"thu\":[{\"start\": \"06:20\",\"end\": \"07:30\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"19:00\",\"end\": \"20:30\",\"mode\": \"on\",\"position\":2}],\"fri\":[{\"start\": \"06:20\",\"end\": \"07:30\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"20:00\",\"end\": \"21:30\",\"mode\": \"on\",\"position\":2}],\"sat\":[{\"start\": \"08:30\",\"end\": \"09:30\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"20:00\",\"end\": \"20:30\",\"mode\": \"on\",\"position\":2},{\"start\": \"20:30\",\"end\": \"21:30\",\"mode\": \"on\",\"position\":3}],\"sun\":[{\"start\": \"09:00\",\"end\": \"10:00\",\"mode\": \"on\",\"position\":0},{\"start\": \"13:30\",\"end\": \"14:30\",\"mode\": \"on\",\"position\":1},{\"start\": \"19:00\",\"end\": \"21:30\",\"mode\": \"on\",\"position\":2}]}" 

$viessmannApi->setRawDhwSchedule($schedule);

I receive:
Message:
         Unable to set data for featureheating.circuits.0.dhw.schedule and action setSchedule and data{"newSchedule": {....
_submitted data_
        Reason: DEVICE_COMMUNICATION_ERROR
Code: 1
`
From VCare app submission of scheduler work fine.

My heater and OPT1 module have functions for eco, comfort and scheduler ex:

- heating.circuits.0.operating.programs.eco
- heating.dhw.schedule

**Comment by thetrueavatar** [comment]
Unfortunately, sometimes we get this message from viessmann's server. There
is nothing I can do because it seems that Viessman server consider there is
a communication problem with the heating device. However, I have notice
this sometimes occurs when everything is fine communication wise and the
error vanish after a couple of hour.
Regards,

**Comment by thetrueavatar** [comment]
Hum for the scheduling problem I have the same problem. My guess it's more
a data format problem than a communication issue.
Same for your new functionality. The best bet to get the right format is to
sniff the data send to Vicare. I used to sniff with packet capture on my
smartphone which does a mitm vpn.
Regards,

**Comment by thetrueavatar** [comment]
Ok I took a close look at this by sniffing through ViCare and there was a
bug. The name of the feature was wrong.
Feature's name is simple heating.dhw.schedule without any circuit
reference. I have fixed this in the develop branch and it should work.
Could you check this ? For your new feature I can't help you since I don't
have that feature on my heating system and can't sniff then the Vicare
data. This is probably the same kind of error(bad data format).
From a REST point of view Viessman's server should have return an HTTP
400(bad request) error but I have already noticed Viessman dev aren't very
concern about error handling...
Thank you anyway for your return.
Best Regards,

**Comment by dr-boss** [comment]
Last night I found he same bug, after post the ex of future  heating.dhw.schedule in message to You when I go sleep ;-))) and when I tested  my modification in your source to send you information  that it work I saw that you found the same bug and you published new version with it correction. 
For eco and comfort program after analyzed sniffed ViCare data and api examples I things that Vissmann block api for change programs remotely.
If I find solution for eco and comfort program for Vissmann with OPT1 module I will send You info.
Thanks and merci for help.
 Piotr

**Comment by thetrueavatar** [comment]
Could you send me your sniffed data packet ?  I have spoofed the identityId
of ViCare so, imho, Viessmann is not able to make the difference between
ViCare and my API.
Moreover, comfort programme from my api can be activated for only one
heater by providing the circuitId. You should have several circuitId if you
have several heater.

public function activateComfortProgram($temperature = NULL, $circuitId = NULL)

Have you already tried this call with a circuitId(1 is defaullt)?
BTW, you can increase "log level" by uncommenting the following line in the
bootstrap.php:

//    echo "Trace:" . $e->getTraceAsString() . "\n";

Regards,

**Comment by dr-boss** [comment]
In my case your code work on my router and communicated directly by WAN port with V.server. I will sniff communication on WAN next week when I have more time or I will install code on machine in LA N.  
The result after uncomments line:
`Message:
         Unable to set data for featureheating.circuits.0.operating.programs.comfort and action activate and data{"temperature":23}
         Reason: DEVICE_COMMUNICATION_ERROR
Code: 1
Trace:#0 phar:///tmp/mnt/ASUS_SYS/entware/share/nginx/html/Viessmann/example/Viessmann-Api-1.2.0-SNAPSHOT.phar/src/API/ViessmannAPI.php(282): Viessmann\API\ViessmannAPI->setRawJsonData('heating.circuit...', 'activate', '{"temperature":...')
#1 /tmp/mnt/ASUS_SYS/entware/share/nginx/html/Viessmann/example/ActivateComfortProgram.php(3): Viessmann\API\ViessmannAPI->activateComfortProgram('23', 0)
#2 {main}`

BTW: I sniffed old version of VCare with wizard for comfort program (wind button on right down corner). And they not start exactly comfort program (stay in home). It change mode to forcenormal to stop time program and launch heating and DHW continuously. For stop  it return to heating&DHW mode. The problem is if you use only DHW mode before launch by Vicare "comfort program". I thing this is reasons why Viessmann removed wizard from ViCare.

**Comment by thetrueavatar** [comment]
Sorry I didn't had the time to check this and forget about it. Do you still having the problem ?

**Comment by JueBag** [comment]
I have a question regarding the message:
```
Unable to set data for featureheating.circuits.0.dhw.schedule and action setSchedule and data{"newSchedule": {....
submitted data
Reason: DEVICE_COMMUNICATION_ERROR
Code: 1
```
IMHO this is coming from the function "setRawJsonData" only. I do see the same message when using the function "startOneTimeDhwCharge". It reports unable but it is performing the desired task. I might missinterpret the code in "setRawJsonData", but this:
`if (isset($response["statusCode"])) { ....`
seems to be treated as TRUE (I can't say due to lacking php knowledge!).

@dr-boss : Confirm you do see this message and the desired command is NOT performed in your case?

[Edit:] I had a wrong function name in that posting. "setRawDhwSchedule" has been changed to te correct "setRawJsonData". Sorry for that!

**Comment by thetrueavatar** [comment]
Which version are you using ? I did a fix recently on this
setRawDhwSchedule because it wasn't working.

Le dim. 1 déc. 2019 à 11:58, Jürgen Baginski <notifications@github.com> a
écrit :

**Comment by thetrueavatar** [comment]
In fact I have change the code from

const DHW_SCHEDULE = "dhw.schedule";

...

public function setRawDhwSchedule($schedule, $circuitId = NULL)
{
    $data = "{\"newSchedule\": $schedule}";
    $this->setRawJsonData($this->buildFeature($circuitId,
self::DHW_SCHEDULE), "setSchedule", $data);
}

to this:

const DHW_SCHEDULE = "heating.dhw.schedule";

public function setRawDhwSchedule($schedule, $circuitId = NULL)

{
    $data = "{\"newSchedule\": $schedule}";
    $this->setRawJsonData(self::DHW_SCHEDULE, "setSchedule", $data);
}

This means, I have removed the possibility to set schedule on a given
circuitId. It wasn't working when I test in on september while using
circuitId but maybe I should keep both. Could you try with the previous
version of the code ?
Regards
Le dim. 1 déc. 2019 à 12:21, thetrueavatar <thetrueavatar@gmail.com> a
écrit :

**Comment by JueBag** [comment]
Ups, I meant to talk about the function "setRawJsonData". I'm sorry for this false posting!
It should read like:
`IMHO this is coming from the function "setRawJsonData" only.`..` I might missinterpret the code in "setRawJsonData", but this:`

**Comment by thetrueavatar** [comment]
DEVICE_COMMUNICATION_ERROR is an message coming from the viessmann's
server. Unfortunately, it use in different case:

- temporary problem with viessman's server which will be resolved by itself

- problems with the format of the data provide when doing a set

The later one was return for startOneTimeDhwCharge and
stopOneTimeDhwCharge and JueBag discover quite recently it was du to
json data provided(should be "empty" {} instead of
{mode:activate/deactivate}.

This has been fixed in 1.2.1 and 1.3.0-SNAPSHOT.

Regards,




Le dim. 1 déc. 2019 à 12:34, Jürgen Baginski <notifications@github.com> a
écrit :

**Comment by dr-boss** [comment]
> Sorry I didn't had the time to check this and forget about it. Do you still having the problem ?

No problem, Your work is big without solved my small problem :-) Thanks for solution 
I use another solution. I add temporary "program nb 4" to DHW schedule from time of set and end after 1h (by retrieve DHW time table, send modified and after 1h return to backup of original). And after this I remove it. This is my party program version for heating water for use

**Comment by Bercolly** [comment]
Bonjour,

Pour un souci de consistance de l'APi et des des méthodes, il serait bon que la méthode :
     public function getHotWaterStorageTemperature($circuit = NULL): string

ait comme paramètre $circuitId plutôt que $circuit.

Cordialement,
Bernard

**Comment by thetrueavatar** [comment]
Merci pour cette remarque pertinente. Je corrigerai ça dès que possible.
Bien à vous,

**Comment by dr-boss** [comment]
> Sorry I didn't had the time to check this and forget about it. Do you still having the problem ?

I understated that I have more important issue and  I doing program with "work around" that is ok for me. If any person request party mode by my home automation system the program added 4th DHW time program from now to 1h plus. And after this one hour the time program was deleted from time table. For me is ok because I never use 4 programs by any day in week. Of course I haven;t any answer from Viessman support, I wrote that in manual and specification is written ... for system like my vCare support remote party and eco mode :-(
